content/posts/exception-handling.md

```diff
@@ -0,0 +1,332 @@
+others' state. 
+
+To ensure that the C fields have no layout conflicts, Python requires the
+following conditions:
+
+1. Every class must have a unique ancestor with the largest `tp_base`. Call this
+   ancestor the _solid base_ of the class.
+2. To multiply inherit from a list of classes `Base1`, `Base2`, ..., `Basen`
+   there must be a class `B` in the list with the property that the solid base
+   of `B` is a subclass of the solid base of every class in the list.
+
+We wish to inherit from both `BaseException` and `BaseJsProxy`. The solid base of
+`BaseException` is `BaseException`, and the solid base of `JsProxy` is `BaseJsProxy`
+but neither `BaseException` nor `BaseJsProxy` is a subclass of the other. So we're
+in trouble.
+
+The first problem we need to solve is the actual layout conflicts. We do this by
+reserving space for the `BaseException` fields in our struct layout:
+
+```C
+struct ExceptionFields
+{
+  PyObject* args;
+  PyObject* notes;
+  // ... there are more of them
+};
+
+typedef struct
+{
+  PyObject_HEAD
+  union {
+    struct ExceptionFields ef;
+    // If the object is not an Exception we can use this space for other stuff.
+    ...
+  } union_fields;
+  // shared fields that are needed by every variant
+  ...
+} BaseJsProxy;
+```
+Now we want to make sure that we've actually laid out the fields correctly, so
+that we get compilation errors rather than segfaults when switching to a new
+Python version that has altered the layout. We use `_Static_assert` for this.
+We check that we have the correct total size:
+```C
+_Static_assert(
+    sizeof(PyObject) + sizeof(struct ExceptionFields) 
+        == sizeof(PyBaseExceptionObject),
+    "size conflict between BaseJsProxy and PyExc_BaseException"
+);
+```
+And that each field has the correct offset:
+```C
+_Static_assert(                                                              
+    offsetof(BaseJsProxy, union_fields) + offsetof(struct ExceptionFields, args) 
+        == offsetof(PyBaseExceptionObject, args),       
+    "layout conflict between BaseJsProxy and PyExc_BaseException");
+```
+Now we know that our `BaseJsProxy` layout matches the layout of `BaseException` and
+it's valid to multiply inherit. However, Python does not know this. In order to
+construct the subclass, we lie to Python about the subclasses of `BaseJsProxy`:
+```C
+// We want our base classes to be BaseJsProxy and Exception
+bases = PyTuple_Pack(2, &BaseJsProxyType, PyExc_Exception);
+PyObject* save_mro = BaseJsProxyType.tp_mro;
+// Tell Python that BaseJsProxy is a subclass of BaseException to convince the
+// class layout conflict detection logic to accept it
+BaseJsProxyType.tp_mro = PyTuple_Pack(1, PyExc_BaseException);
+// create our JsException type
+JsExceptionType = PyType_FromSpecWithBases(&spec, bases);
+// Put the mro back before anyone notices!
+Py_CLEAR(BaseJsProxyType.tp_mro);
+BaseJsProxyType.tp_mro = save_mro;
+```
+
+## Conclusion
+
+??? What should we say here?
```

Comment:
Maybe adding some future plans or briefly mentioning some weird cases that are not handled in this post (c++ exceptions?) would be nice?